### PR TITLE
Openai Upstash caption generation

### PR DESCRIPTION
**Pull request by andrelandgraf**
## Summary (1-2 sentences)

This PR integrates Upstash QStash for background task management and uses OpenAI to generate group image alt texts in a background task.

## Details

This PR introduces two new dependencies:

- [QStash (Upstash)](https://upstash.com/docs/qstash/overall/getstarted)
- OpenAI

### QSash

QStash is a message queue-ish SaaS from Upstash that manages background tasks for us. We can send QStash a POST request with a URL and request body, and Upstash will make sure to call that URL with the body content later (in the background). If the request fails, Upstash will document the failed task on their dashboard and try to re-trigger it later. Our application still runs the background task, but Upstash takes over managing retries, logging, and scheduling.

### OpenAI

GPT endpoint to generate text content based on text and image input.

### CODE REVIEW

File: app/routes/api.generate-image-caption.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import type { ActionFunctionArgs } from '@remix-run/node';
```
Comment (shami-dev) [inline app/routes/api.generate-image-caption.ts:1]: Probably, I missed the idea. What is the purpose of calling this route starting with api? Why should it be in the routes folder?

File: app/routes/groups_.new.tsx [hunk 2]
```diff
@@ -45,11 +46,17 @@ export async function action({ request }: ActionFunctionArgs) {
     groupImageUrl = `https://fzzehiiwadkmbvpouotf.supabase.co/storage/v1/object/public/group-cover-images/${data.path}`;
   }
 
-  // TODO generate alt text or prompt user for alt text
   const new_group = await db.group.create({
-    data: { name, description, imgUrl: groupImageUrl, imgAlt: 'Group image' },
+    data: { name, description, imgUrl: groupImageUrl, imgAlt: 'The group image' },
```
Comment (shami-dev) [inline app/routes/groups_.new.tsx:50]: Would placing a group name by default instead of 'The group image' be better?

**Review by shami-dev — COMMENTED** [review]
Great implementation and learning! 
Thank you! 🙏

File: app/modules/gpt/generate.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    content.push({
+      type: 'image_url',
+      image_url: {
+        url: imageUrl,
+        detail: 'auto',
+      },
+    });
+  }
+  const response = await openai.chat.completions.create({
+    model: 'gpt-4-o',
```
[the remark this answers is not in the document]
Comment (andrelandgraf) [reply app/modules/gpt/generate.ts:27]: Great catch! Fixed!

File: app/routes/groups_.new.tsx (hunk 2 above)
Comment (andrelandgraf) [reply app/routes/groups_.new.tsx:50]: Not sure! The name of the group will most likely already be a heading on the page, so I believe adding it to the image alt text group be repetitive?

**Review by eiffelwong1 — APPROVED** [review]
LGTM

**Review by tonydangblog — APPROVED** [review]

**Pull request merged**